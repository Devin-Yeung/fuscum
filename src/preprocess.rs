use std::borrow::Cow;

pub mod python;

pub trait Preprocessor {
    fn preprocess<S: AsRef<str>>(src: &S) -> Cow<str>;
}
