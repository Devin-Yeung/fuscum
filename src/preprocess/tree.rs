use ast_grep_core::matcher::KindMatcher;
use ast_grep_core::source::Edit;
use ast_grep_core::{AstGrep, Language, StrDoc};

pub struct Tree<L: Language> {
    lang: L,
    ag: AstGrep<StrDoc<L>>,
}

impl<L: Language + Copy> Tree<L> {
    pub fn new<S: AsRef<str>>(src: S, lang: L) -> Self {
        Self {
            ag: lang.ast_grep(src.as_ref()),
            lang,
        }
    }

    pub fn apply_edit_helper(&mut self, edits: Vec<Edit<String>>) -> String {
        debug_assert_ne!(edits.len(), 0);
        let mut new_content = String::new();
        let old_content = self.ag.root().root().get_text();
        let mut start = 0;
        for diff in edits {
            let range = diff.position..diff.position + diff.deleted_length;
            new_content.push_str(&old_content[start..range.start]);
            let replacement = String::from_utf8(diff.inserted_text).unwrap();
            new_content.push_str(&replacement);
            start = range.end;
        }
        // add trailing statements
        new_content.push_str(&old_content[start..]);
        new_content
    }

    pub fn apply_edits(&mut self, edits: Vec<Edit<String>>) -> &mut Self {
        match edits.len() {
            0 => self,
            _ => {
                let new_content = self.apply_edit_helper(edits);
                self.ag = self.lang.ast_grep(&new_content);
                self
            }
        }
    }

    pub fn remove_comments<S: AsRef<str>>(&mut self, kind: S) -> &mut Self {
        let pat = KindMatcher::new(kind.as_ref(), self.lang);
        let edits = self.ag.root().replace_all(&pat, "");
        self.apply_edits(edits)
    }

    pub fn subst_ident<S: AsRef<str>>(&mut self, kind: S, to: &str) -> &mut Self {
        let pat = KindMatcher::new(kind.as_ref(), self.lang);
        let edits = self.ag.root().replace_all(&pat, to);
        self.apply_edits(edits)
    }

    pub fn subst_string<S: AsRef<str>>(&mut self, kind: S, to: &str) -> &mut Self {
        let pat = KindMatcher::new(kind.as_ref(), self.lang);
        let edits = self.ag.root().replace_all(&pat, to);
        self.apply_edits(edits)
    }

    pub fn source(&self) -> &str {
        self.ag.source()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ast_grep_language::Python;

    #[test]
    fn var_subst() {
        let mut tree = Tree::new("def f(a, b, c):\n\ta = 1", Python);
        tree.subst_ident("identifier", "v");
        assert_eq!(tree.source(), "def v(v, v, v):\n\tv = 1");
    }
}
