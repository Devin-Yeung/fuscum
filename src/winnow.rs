use std::hash::{DefaultHasher, Hash, Hasher};

pub fn k_gram(text: &str, k: usize) -> Vec<u64> {
    let mut hashes = Vec::new();

    for i in 0..text.len().saturating_sub(k) {
        let kgram = &text[i..i + k];
        let mut hasher = DefaultHasher::new();
        kgram.hash(&mut hasher);
        hashes.push(hasher.finish());
    }

    hashes
}

pub fn winnowing<T>(hashes: T, window_size: usize) -> Vec<(u64, usize)>
where
    T: AsRef<[u64]>,
{
    let seq = hashes.as_ref();
    let mut finger_prints = Vec::new();
    // the global position of the minimum hash value in the *previous* window
    let mut idx = 0usize;

    for i in 0..seq.len().saturating_sub(window_size) {
        let window = &seq[i..i + window_size];
        let (min_hash, min_idx) = rightmost_minimal(window);
        if idx != i + min_idx {
            // make sure it's not the same as the last window
            idx = i + min_idx;
            finger_prints.push((min_hash, idx));
        }
    }

    finger_prints
}

/// Find the rightmost minimal hash value in the window
fn rightmost_minimal(window: &[u64]) -> (u64, usize) {
    let mut min = u64::MAX;
    let mut idx = 0usize;
    for (i, &hash) in window.iter().enumerate() {
        if hash < min {
            min = hash;
            idx = i;
        }
    }
    (min, idx)
}

#[cfg(test)]
mod tests {
    use crate::winnow::k_gram;

    #[test]
    fn it_works() {
        let hashes = [
            77u64, 74, 42, 17, 98, 50, 17, 98, 8, 88, 67, 39, 77, 74, 42, 17, 98,
        ];
        let window_size = 4;
        assert_eq!(
            super::winnowing(hashes, window_size),
            // [(17, 3), (17, 6), (8, 8), (39, 11), (17, 15)]
            vec![(17, 3), (17, 6), (8, 8), (39, 11), (17, 15)]
        );
    }

    #[test]
    fn k_gram_works() {
        let text = "adorunrunrunadorunrun";
        let k = 5;
        assert_eq!(
            k_gram(text, k),
            vec![
                7536710649711940037,
                12375835004367686960,
                13240722851591535538,
                4020085029674966483,
                4972485008023615292,
                1468765096528618582,
                4020085029674966483,
                4972485008023615292,
                2165872647979677269,
                2880295526655702587,
                9732345111308041966,
                13607179090089924327,
                7536710649711940037,
                12375835004367686960,
                13240722851591535538,
                4020085029674966483
            ]
        );
    }
}
