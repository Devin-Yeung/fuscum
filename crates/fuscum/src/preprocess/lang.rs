use crate::preprocess::tree::Tree;
use crate::preprocess::Preprocessor;
use std::borrow::Cow;
use typed_builder::TypedBuilder;

macro_rules! impl_lang_preprocessor {
    (
        $name:ident,
        lang => $lang:expr,
        ident => $identifier_token:expr,
        string => $string_token:expr,
        comment => $comment_token:expr
    ) => {
        #[derive(TypedBuilder)]
        #[builder(doc, field_defaults(default, setter(into)))]
        pub struct $name {
            #[builder(setter(
                strip_option,
                doc = "Substitute variable to a certain unified representation"
            ))]
            subst_var: Option<String>,
            #[builder(setter(
                strip_option,
                doc = "Substitute string to a certain unified representation"
            ))]
            subst_string: Option<String>,
            #[builder(default = true, setter(doc = "Remove comments from the source code"))]
            remove_comments: bool,
        }

        impl Default for $name {
            fn default() -> Self {
                Self {
                    subst_var: Some("v".to_string()),
                    subst_string: Some("\"s\"".to_string()),
                    remove_comments: true,
                }
            }
        }

        impl Preprocessor for $name {
            fn preprocess(&self, src: &str) -> Cow<str> {
                let mut tree = Tree::new(src, $lang);

                if self.remove_comments {
                    tree.remove_comments($comment_token);
                }

                self.subst_var
                    .as_ref()
                    .map(|v| tree.subst_ident($identifier_token, v));

                self.subst_string
                    .as_ref()
                    .map(|v| tree.subst_string($string_token, v));

                let src = tree
                    .source()
                    .chars()
                    .filter(|c| !c.is_whitespace())
                    .collect::<String>();
                Cow::Owned(src)
            }
        }
    };
}

impl_lang_preprocessor!(
    PythonPreprocessor,
    lang => ast_grep_language::Python,
    ident => "identifier",
    string => "string",
    comment => "comment"
);

impl_lang_preprocessor!(
    CPreprocessor,
    lang => ast_grep_language::C,
    ident => "identifier",
    string => "string_literal",
    comment => "comment"
);

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn python() {
        let src = include_str!("../../../../fixtures/langs/python.py");
        let pp = PythonPreprocessor::default();
        let res = pp.preprocess(&src);
        insta::assert_snapshot!(&format!("{src}\n\n\n{res}"))
    }

    #[test]
    fn c() {
        let src = include_str!("../../../../fixtures/langs/c.c");
        let pp = CPreprocessor::default();
        let res = pp.preprocess(&src);
        insta::assert_snapshot!(&format!("{src}\n\n\n{res}"))
    }
}
