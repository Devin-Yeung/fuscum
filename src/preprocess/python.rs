use crate::preprocess::Preprocessor;
use ast_grep_core::matcher::KindMatcher;
use ast_grep_core::source::Edit;
use ast_grep_core::{AstGrep, Language, Pattern, StrDoc};
use ast_grep_language::Python;
use std::borrow::Cow;

pub struct PythonPreprocessor;
pub struct PyTree {
    ag: AstGrep<StrDoc<Python>>,
}

impl PyTree {
    pub fn new<S: AsRef<str>>(src: S) -> Self {
        Self {
            ag: Python.ast_grep(src.as_ref()),
        }
    }

    pub fn apply_edits(&mut self, edits: Vec<Edit<String>>) -> &mut Self {
        edits
            .into_iter()
            .rev()
            .fold(&mut self.ag, |root, edit| root.edit(edit).unwrap());
        self
    }

    pub fn remove_comments(&mut self) -> &mut Self {
        let pat = KindMatcher::new("comment", Python);
        let edits = self.ag.root().replace_all(&pat, "");
        self.apply_edits(edits)
    }

    pub fn subst_ident(&mut self, to: &str) -> &mut Self {
        let pat = KindMatcher::new("identifier", Python);
        let edits = self.ag.root().replace_all(&pat, to);
        self.apply_edits(edits) // TODO: bottleneck, it's too slow
    }

    pub fn source(&self) -> &str {
        self.ag.source()
    }
}

impl Preprocessor for PythonPreprocessor {
    fn preprocess<S: AsRef<str>>(&self, src: &S) -> Cow<str> {
        let mut tree = PyTree::new(src);
        tree.subst_ident("v").remove_comments();
        // remove all the whitespace in the source code
        let src = tree
            .source()
            .chars()
            .filter(|c| !c.is_whitespace())
            .collect::<String>();
        Cow::Owned(src)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn var_subst() {
        let mut tree = PyTree::new("def f(a, b, c):\n\ta = 1");
        tree.subst_ident("v");
        assert_eq!(tree.source(), "def v(v, v, v):\n\tv = 1");
    }

    #[test]
    fn preprocess() {
        let src = "def f(a, b, c):\n\ta = 1";
        let pp = PythonPreprocessor.preprocess(&src);
        assert_eq!(pp, "defv(v,v,v):v=1");
    }
}
