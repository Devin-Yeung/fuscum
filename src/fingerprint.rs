use crate::preprocess::Preprocessor;
use crate::winnow::{k_gram, winnowing};

pub struct FingerPrintConfig {
    pub k: usize,
    pub window_size: usize,
}

impl FingerPrintConfig {
    pub fn new(k: usize, window_size: usize) -> Self {
        Self { k, window_size }
    }
}

impl Default for FingerPrintConfig {
    fn default() -> Self {
        Self {
            k: 35,
            window_size: 5,
        }
    }
}

pub struct FingerPrint {
    fingerprints: Vec<(u64, usize)>,
}

impl FingerPrint {
    pub fn new<P, S>(src: S, config: FingerPrintConfig) -> Self
    where
        P: Preprocessor,
        S: AsRef<str>,
    {
        let src = P::preprocess(&src);
        let hashes = k_gram(&src, config.k);
        let fingerprints = winnowing(&hashes, config.window_size);
        Self { fingerprints }
    }

    pub fn fingerprints(&self) -> &[(u64, usize)] {
        &self.fingerprints
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::preprocess::python;

    #[test]
    fn it_works() {
        let src = "def f(a, b, c):\n\ta = 1";
        let fp =
            FingerPrint::new::<python::PythonPreprocessor, _>(src, FingerPrintConfig::new(3, 3));
        insta::assert_debug_snapshot!(fp.fingerprints());
    }
}
