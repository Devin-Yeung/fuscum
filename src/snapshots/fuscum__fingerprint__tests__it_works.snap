---
source: src/fingerprint.rs
expression: fp.fingerprints()
snapshot_kind: text
---
[
    (
        5280567008640046432,
        2,
    ),
    (
        4341634741185337740,
        4,
    ),
    (
        8486556841783744955,
        6,
    ),
    (
        74726906490296853,
        8,
    ),
]
